### ISSUE TITLE
Attestation for policy verification

### ISSUE THREAD

**Issue by kpk47** [in-toto/attestation#277]
[SLSA](https://slsa.dev) has the concept of a Verification Summary Attestation ([VSA](https://slsa.dev/spec/v1.0/verification_summary)). VSA is an in-toto attestation that indicates the attestor verified an artifact's SLSA metadata and found it to meet a certain SLSA level. SLSA users can use VSAs to cache verification decisions and/or delegate such decisions to another trusted party. VSAs are specific to SLSA, but perhaps it would be useful to have an attestation that was agnostic to the type of policy the attestor used. That attestation would mean “<verifier> evaluated <subject> against <policy_uri> and it (passed|failed).” We've discussed creating such an attestation as part of SLSA, but the community agreed that it was outside SLSA's scope. I'm curious whether in-toto would be interested in pursuing the idea.

IIUC, such an attestation is slightly outside of in-toto’s usual scope since it’s a statement about the attestor’s beliefs about the subject rather than metadata about the subject itself, but the framing allows us to move policy decisions further up the supply chain and away from admission control. For example, instead of gathering up a bundle of attestations for SLSA, vuln scans, etc and evaluating a policy about those attestations during admission control, you could check that policy into version control, evaluate it and produce a verification attestation, and then at admission control you have a much simpler policy that checks that there’s a passing verification attestation for the subject that uses the correct policy and comes from a trusted verifier.

As far as I can tell, in-toto doesn’t support inheritance, but I could imagine different sorts of inheritance attestations over time. The essential elements are the identity of the verifier, a policy URI, and whether the verification passed or failed. I’m not sure what to do about divergence between different verification attestations or whether it’s premature to think about now.

**Comment by JustinCappos** [comment]
I don't think this is out of scope.  It's certainly possible to make an attestation that some process passed or failed and have it relate to that functionary's beliefs.  From my perspective this is something we already support.  

Do you feel that something is missing from our codebase or spec?  Is there a reason this cannot be done off-the-shelf today?  Is it a documentation issue instead?

(I'm punting on the inheritance aspects of your suggestion though.)

**Comment by kpk47** [comment]
Maybe this is a documentation issue? I think I'm mainly looking for validation of the idea and feedback on whether it's in or out of in-toto's scope. Inheritance is the hard part, so I suppose I'm implicitly asking if there's appetite for a change to an attestation layer or adding a new layer. Would it be better if I submitted a proposal for a new predicate?

cc @TomHennen

**Comment by kpk47** [comment]
There may also be a documentation problem, as all the examples I found discuss applying in-toto policies at admission control time. If in-toto supports moving policy checks earlier, then perhaps there could be an example of that pattern.

**Comment by adityasaky** [comment]
> IIUC, such an attestation is slightly outside of in-toto’s usual scope since it’s a statement about the attestor’s beliefs about the subject rather than metadata about the subject itself

This is an interesting distinction. In my mental model, I apply that filter for all attestations as we always have to take some claims at face value as stated in the attestation.

> For example, instead of gathering up a bundle of attestations for SLSA, vuln scans, etc and evaluating a policy about those attestations during admission control, you could check that policy into version control, evaluate it and produce a verification attestation, and then at admission control you have a much simpler policy that checks that there’s a passing verification attestation for the subject that uses the correct policy and comes from a trusted verifier.

IMO this is very much in-scope for in-toto. The verification workflow in in-toto's specification is recursive, resulting in an attestation (and in fact we're now exploring how we can use VSAs here) that is then verified by a "higher level" policy. This is commonly seen in the sublayout semantic but we've explored using verification summaries for low-resource contexts such as automotive software.

> As far as I can tell, in-toto doesn’t support inheritance, but I could imagine different sorts of inheritance attestations over time. The essential elements are the identity of the verifier, a policy URI, and whether the verification passed or failed. I’m not sure what to do about divergence between different verification attestations or whether it’s premature to think about now.

The word "inheritance" is throwing me here (we've briefly thrown some inheritance-like ideas for predicates to complicate matters in my mind) but if my read of what you mean by it is right, I'm quite supportive of this.

> There may also be a documentation problem, as all the examples I found discuss applying in-toto policies at admission control time. If in-toto supports moving policy checks earlier, then perhaps there could be an example of that pattern.

I don't know if I can speak for the attestation maintainers / other in-toto spec authors but for me "admission controller" is quite generic. A dependency review attestation (#151) would be verified prior to a library being admitted to the source control system, for example. What I'm getting at is that in my mind we _already_ have this notion of verification at multiple points as demanded by context but perhaps it's not universal? If others have the same view, we should definitely make that more explicit in the docs.

**Comment by TomHennen** [comment]
Yeah I'd be supportive of something like this.

Is inheritance strictly necessary?  I suspect we can probably meet existing needs without it, but maybe you have something in mind that won't work without it?

I think one thing that we'd discussed is 'moving' the VSA from SLSA to in-toto proper and making it more generic when we do so.  If people want to attest to various different properties that were checked they could just go in what is now the `verifiedLevels` list, but could be renamed something like `propertiesVerified`.

**Comment by kpk47** [comment]
I think "inheritance" may be the wrong word. I'm thinking of something more like [extensions](https://protobuf.dev/programming-guides/proto2/#extensions) in Protocol Buffers. Users could optionally add fields to the attestation to give supplemental information about the policy that they verified. Though the more I think about it, the more I think that such information would be better put in a separate attestation that's part of the same bundle. Would that be more idiomatic to in-toto?

> I think one thing that we'd discussed is 'moving' the VSA from SLSA to in-toto proper and making it more generic when we do so. If people want to attest to various different properties that were checked they could just go in what is now the `verifiedLevels` list, but could be renamed something like `propertiesVerified`.

Thank you for bringing this up, and I apologize for not including the context. If in-toto had a "policy verification" attestation, then the SLSA community may encourage users to use that attestation instead of or in addition to VSAs.

**Comment by TomHennen** [comment]
Chatted about this at the maintainers meeting today.

1. We're generally for moving the VSA from SLSA to in-toto (and generalizing it in the process)
2. SLSA specific things, if needed, could be added via annotations of some sort.
3. We'd probably want the outstanding VSA PRs in the SLSA repo to be resolved one way or another before moving.
4. We'd want to avoid there being multiple different ways to represent the same thing (e.g. let's not have VSAs and something else that's very close to VSAs if we can avoid it).

**Comment by mikhailswift** [comment]
This is very relevant to our current work on Witness: https://github.com/testifysec/witness/issues/14 . We've had this issue in our backlog but are circling back to it in the short term.

After a quick read through the VSA spec, I think it would fit Witness's needs well. The only caveats are with number 2 in Tom's comment. Witness isn't currently asserting an artifact's SLSA level, though it may be able to in the future.

**Comment by adityasaky** [comment]
Indeed, but that work needs some more clarity on the status of VSA as a generic predicate type. https://github.com/in-toto/attestation-verifier is a great place to test out how VSAs can work for ITE-10/11 layouts.

**Comment by kpk47** [comment]
Glad to hear there's interest in moving VSA to in-toto. I'm working on closing out the remaining VSA PRs on the SLSA repo and should have that done shortly.

**Comment by TomHennen** [comment]
We've had some discussions about this internally and I put this together to share those thoughts here in case it helps: https://docs.google.com/presentation/d/1_1ikfSaja87-Qx2ev26DYdiNgS6AEvVFO92oSwNGvJ8/edit?usp=sharing  Comments welcome.

**Comment by arewm** [comment]
This topic came up in the SLSA specification call today. Namely, the question was raised whether in-toto might need a new attestation type or whether [SCAI](https://github.com/in-toto/attestation/blob/main/spec/predicates/scai.md) would be able to be adapted to fulfil the needs of the SLSA verification summary attestations.

After reading the specification, it SCAI seems to be focused only on providing additional information and evidence (i.e. to potentially support future policy verification actions). @marcelamelara, do you have any additional context/proposal for how SCAI might be able to fit the needs of a VSA use case?

**Comment by marcelamelara** [comment]
Based on Tom's slide deck, I've seen a couple of parallels between the fields in the [simplified VSA format](https://docs.google.com/presentation/d/1_1ikfSaja87-Qx2ev26DYdiNgS6AEvVFO92oSwNGvJ8/edit#slide=id.g1f2bfdcb431_0_321) and the SCAI predicate:
*  `verifier` <> `producer` in SCAI
*  `properties` <> `attributes` in SCAI, where each verified attestation could be referenced as the `target` for a particular individual `attribute`. And since SCAI doesn't require the `evidence` field, it could be left empty.
 
The `timeVerified` field, or something similar, could possibly be added as an optional field to the SCAI spec. That said, the SCAI predicate, as currently specified, is quite generic, so there's definitely a need to dig a bit deeper into whether it provides all the needed semantics from a policy verification standpoint.

My plan for validating if/how SCAI could fit the needs of a VSA is to develop some examples for the types of policy checks/results that we want to be able to express. I can think of a few examples, but I definitely want to involve the community in this process so we can make sure we're covering a broader set of policy verification use cases.

**Comment by adityasaky** [comment]
(It's a bit late but I hope this comment is fairly clear...)

I recognize there's a lot of overlap in that VSA and SCAI both capture some set of (arbitrary-ish) attributes but I'm not sure that's necessarily enough to merge them into a single type. In my mind, there's a lot of value in seeing a VSA predicate type with the expectation that some verification in fact happened. While this can be communicated using SCAI, clients may have to do more work to distinguish a generic SCAI attestation and one specifically for policy verification.

I'll also highlight that SCAI is incredibly powerful precisely _because it is generic and can be used for a variety of use cases_, so I imagine we could argue for using SCAI in place of several other predicates, not just the VSA. Whether we should, IMO, goes back to the philosophy of having types in the first place to communicate something more specific.

**Comment by marcelamelara** [comment]
> Whether we should, IMO, goes back to the philosophy of having types in the first place to communicate something more specific with each one.

I completely agree with this. The main issue with using SCAI out of the box is that the predicateType doesn't convey the semantics. So, I've been thinking about whether we could introduce a way to sub-class existing predicates in cases like these. I can see this being useful not only in this particular scenario, but others as well. This idea probably veers into new ITE territory, so it may not be as immediately actionable.

**Comment by adityasaky** [comment]
> So, I've been thinking about whether we could introduce a way to sub-class existing predicates in cases like these.

Without going too far off topic, I think this has come up before for sub-typing predicates, specifically for human review types. I could see this for a generic SBOM type that's a parent for SPDX and CDX, various subtypes for SCAI, etc. I'm definitely down to explore this more generally as a separate ITE, and what semantics we'd define for sub-typing in the attestation framework!

**Comment by kpk47** [comment]
> The main issue with using SCAI out of the box is that the predicateType doesn't convey the semantics. 

VSA also has that issue, which is part of the reason I stopped working on this PR. Attestations semantics are the hard part, not the format, so I felt like I was expending an undue amount of effort solving the easy part of a problem. I'm hopeful that whatever we end up using will comment on the policy's semantics in addition to whether or not the policy check passed.

> So, I've been thinking about whether we could introduce a way to sub-class existing predicates in cases like these. I can see this being useful not only in this particular scenario, but others as well. This idea probably veers into new ITE territory, so it may not be as immediately actionable.

I also thought about sub-classing, but I couldn't convince myself that it really solved a problem unless it were paired with different expectations for how to define the types. If it's just as hard to create a sub-class of an existing predicate as it is to create a new predicate, why sub-class?

**Comment by laurentsimon** [comment]
> I also thought about sub-classing, but I couldn't convince myself that it really solved a problem unless it were paired with different expectations for how to define the types. If it's just as hard to create a sub-class of an existing predicate as it is to create a new predicate, why sub-class?

+1. Sometimes a little bit of duplication is better than adding a lot of complexity.

**Comment by marcelamelara** [comment]
> Attestations semantics are the hard part, not the format

100% and I appreciate you pointing this out again @kpk47 . 

This brings me back to my earlier point about needing concrete policy examples and requirements. These will hopefully allow us to derive common semantics that consumers of policy verification attestations can reason about, and when we're ready to discuss formats again also find/develop a suitable one.

**Comment by arewm** [comment]
The desire to move the specification out of SLSA came up again at OSS-NA this week. With the SLSA v1.2 track, we are proposing the use of "verified properties" which do not cleanly fit into the current VSA spec. We are also exploring the possibility of using VSAs as a tool for additional tracks. To enable the VSA to continue evolving (and to evolve outside of versions coupled with SLSA), we wanted to again try to propose moving the specification to in-toto.

In talking with @TomHennen this week, the rough proposal is shown [in this deck](https://docs.google.com/presentation/d/1_1ikfSaja87-Qx2ev26DYdiNgS6AEvVFO92oSwNGvJ8/edit?slide=id.g1f2bfdcb431_0_321#slide=id.g1f2bfdcb431_0_321) and repeated below

```json
// Predicate
"predicateType": "https://in-toto.io/attestation/verification_summary/v0.1",
"predicate": {
  "verifier": {
    "id": "<URI>"
    }
  },
  "timeVerified": <TIMESTAMP>,
  "resourceUri": <artifact-URI-in-request>,
  "properties": ["<SCOPED_VERIFIED_PROPERTY>"],
}
```

An example predicate might be:

```
  "predicateType": "https://in-toto.io/attestation/verification_summary/v0.1",
  "predicate": {
    "verifier": {
      "id": "https://example.com/publication_verifier"
    },
    "timeVerified": "1985-04-12T23:20:50.52Z",
    "resourceUri": "https://example.com/example-1.2.3.tar.gz",
    "properties": ["SLSA_BUILD_LEVEL_3", "SLSA_SOURCE_LEVEL_3", "ORG_SOURCE_STATIC_ANALYSIS", "RELEASE_APPROVED", "CONFORMA_HERMETIC_BUILD_TASK"]
  }
```

This is a concise representation of a verification and notable does NOT contain any information for reproducing the verified properties. In order to assist in reproducing the verification result, however, we can recommend to generate a SLSA provenance attestation for the verification summary.

If we want to change the name to more uniquely differentiate between a SLSA VSA, we can also rename this to something like a "simplified_verification" predicate.

@marcelamelara @adityasaky @jkjell @lcarva @AdamZWu @trishankatdatadog @MarkLodato, fya as reviewers to the previous PR.

**Comment by arewm** [comment]
I just created a new PR: #470!

**Comment by TomHennen** [comment]
@AdamZWu 
> For example, a verification result against SLSA build provenance is likely very durable; however, verification results against a vulnerability scanning result attestation have a rather short freshness span.

When we've thought about this before (and I _think_ the way it's been implemented) those time sensitive properties wind up defining some sort of validity period.   So the `VULN_SCANNED` property _could_ be defined as "Has a vulnerability scan attestation no older than 1 day".  Users can then pair that information with the existing `timeVerified` attestation to determine if that's "good enough" or not.

I get the sense that _doesn't_ work for your use case.  Can you say a bit more about why?

**Comment by AdamZWu** [comment]
> [@AdamZWu](https://github.com/AdamZWu)
> 
> When we've thought about this before (and I _think_ the way it's been implemented) those time sensitive properties wind up defining some sort of validity period. So the `VULN_SCANNED` property _could_ be defined as "Has a vulnerability scan attestation no older than 1 day". Users can then pair that information with the existing `timeVerified` attestation to determine if that's "good enough" or not.
> 
> I get the sense that _doesn't_ work for your use case. Can you say a bit more about why?

I think the above scheme *could* work. However, it is a bit too nuanced under the current schema:
 
- Since the `verifiedLevels` is a simple string array, there is no easy way to extend it.
- So we would have to resort to establishing a separate custom extension to hold the data.
- Consider how we keep the records:
   - We could establish an array of objects, and use the index value as the key, something like:
      ```
      "verifiedLevels": [ "SLSA_BUILD_L2", "ATTR_A", "ATTR_B" ]
      "\\foo.org\AttributeTiming": [ {"expiry": null}, {"expiry": "2025-12-12T12:00:00Z" }, {"expiry": "..."} ]
      ```

      However, this feels a bit fragile. Plus it has bad readability -- it is hard to tell which attribute correspond to which timing objects
   - Alternatively, we can use an object, with the attribute directly as the key:
      ```
      "verifiedLevels": [ "SLSA_BUILD_L2", "ATTR_A", "ATTR_B" ]
      "\\foo.org\AttributeTiming": {
        "SLSA_BUILD_L2": {"expiry": null},
        "ATTR_A": {"expiry": "2025-12-12T12:00:00Z" },
        "ATTR_B": {"expiry": "..."}
      }
      ```

      I like this one better. However, the new custom extension is basically a superset of "verifiedLevels", rendering the latter completely redundant.

So, I am hoping SLSA / in-toto could actually officially replace the text array with something similar to the extended object... :P

**Comment by TomHennen** [comment]
Based on our discussions a couple weeks ago that is the sort of complexity that we'd like to avoid if we can.  It would no longer be simple.

The general suggestion is that we _not_ encode expiry (or a generation timestamp) at all, and instead let people infer the times when they need to _or_ have them require an additional attestation if needed.  E.g. if you really need a fresh vulnerability scan, then require the vuln scan itself in addition to the VSA.
